pynndescent/pynndescent_.py

```diff
@@ -161,6 +162,7 @@ def nn_descent(
     seed_per_row=False,
 ):
     n_vertices = data.shape[0]
+    tried = set([(-1, -1)])
```

Comment:
Keeping a record of pairs that have been tried is an optimization, rather than a change in behaviour. @lmcinnes will have more context about why this was added.

It may not work as well in a parallel (threaded) setting where each thread would presumably have its own thread to avoid contention. Also, I wonder if numba can compile this in `nopython` mode.

Reply:
It offers some speedups in initialization. I suspect it will be less effective for the threaded version, but it may be worth benchmarking with and without. It is compatible with nopython mode -- numba is quite happy to compile sets of homogeneous type -- which is why this gets the non-empty initialization: it is effectively a type hint to numba so that it can compile the set.